src/charm.py

```diff
@@ -550,36 +562,28 @@ def _on_upgrade_charm(self, _: UpgradeCharmEvent) -> None:
         """Handle the update of charm."""
         state = self._setup_state()
 
-        logger.info("Reinstalling dependencies...")
-        try:
-            # The `_start_services`, `_install_deps` includes retry.
-            self._install_deps()
-            self._start_services(state.charm_config.token, state.proxy_config)
-            metrics.setup_logrotate()
-        except (LogrotateSetupError, SubprocessError) as err:
-            logger.exception(err)
-
-            if isinstance(err, LogrotateSetupError):
-                msg = "Failed to setup logrotate"
-            else:
-                msg = "Failed to install dependencies"
-            self.unit.status = BlockedStatus(msg)
+        if self._block_on_openstack_config(state):
             return
 
-        state = self._setup_state()
-        self._refresh_firewall(state)
-        logger.info("Flushing the runners...")
+        logger.info("Reinstalling dependencies...")
+        self._common_install_code(state)
+
         runner_manager = self._get_runner_manager(state)
-        if not runner_manager:
-            return
+        runner_manager.schedule_build_runner_image()
 
+        logger.info("Flushing the runners...")
         runner_manager.flush(FlushMode.FLUSH_BUSY_WAIT_REPO_CHECK)
         self._reconcile_runners(
             runner_manager,
             state.runner_config.virtual_machines,
             state.runner_config.virtual_machine_resources,
         )
 
+        # Ensure that an update to the systemd timer for reconciliation event is written to disk.
+        # The upgrade_charm event is not necessarily followed by a config changed event
```

Comment:
Reading here it says the opposite (there is always a config-changed):  https://juju.is/docs/sdk/upgrade-charm-event ```In some cases, [config-changed](https://juju.is/docs/sdk/config-changed-event) can be used instead of install and upgrade-charm because it is guaranteed to fire after both.```

Reply:
Thanks for finding this. This contractdicts my experience (reconciliation timer was not updated after a charm upgrade) and the graph in https://juju.is/docs/sdk/charm-lifecycle does not show this either. We'll have to check with the Juju team.

Reply:
fwiw https://juju.is/docs/sdk/charm-lifecycle has been updated